### PR TITLE
roger - no ticket - add PanelCard component

### PR DESCRIPTION
**Pull request by roggc**
Screenshot 2022-01-17 at 00 59 29
Screenshot 2022-01-17 at 15 38 17

### CODE REVIEW

File: src/Containers/PanelCard/PanelCard.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import React, { useCallback } from 'react';
+import { Button } from '@Inputs/Button/Button';
+import { CardProps, Card } from '../Card/Card';
+
+export enum OperationState {
```
Comment (ralph-dev) [inline src/Containers/PanelCard/PanelCard.tsx:5]: Remove this it's not needed. Let the user pass what they want to pass

File: src/Containers/PanelCard/PanelCard.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    isLoading,
+    isFailure,
+    isSuccess,
+    isUnknown,
+}
+
+export interface IPanelCardProps extends CardProps {
+    messageIsFailure?: string;
+    messageIsSuccess?: string;
+    messageIsLoading?: string;
```
Comment (ralph-dev) [inline src/Containers/PanelCard/PanelCard.tsx:15]: common strings are at the end.

isFailureMessage
isSuccessMessage
isLoadingMessage

File: src/Containers/PanelCard/PanelCard.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import { Meta, Story } from '@storybook/react';
+import { PanelCard, IPanelCardProps, OperationState } from './PanelCard';
+
+export default {
+    title: 'Components/PanelCard',
+    component: PanelCard,
```
Comment (ralph-dev) [inline src/Containers/PanelCard/PanelCard.stories.tsx:7]: argtypes are missing on cancel action()

**Review by ralph-dev — CHANGES_REQUESTED** [review]

File: src/Containers/PanelCard/PanelCard.stories.tsx (hunk 3 above)
Comment (roggc) [reply src/Containers/PanelCard/PanelCard.stories.tsx:7]: @ralph-dev , I don't fully understand argtypes in storybook

Comment (ralph-dev) [reply src/Containers/PanelCard/PanelCard.stories.tsx:7]: Look at Button how it implements actions so you can record button clicks.

ArgTypes is just 

argTypes: {onClickHandler: action("message on click")}

**Comment by ralph-dev** [comment]
Order of priority

error
Loading
Success

Becuase if there's an error who cares if it's loading or success

If there's loading you can be sure it's success

Success is success

**Comment by roggc** [comment]
> Order of priority
> 
> error Loading Success
> 
> Becuase if there's an error who cares if it's loading or success
> 
> If there's loading you can be sure it's success
> 
> Success is success

I don't get it

File: src/Containers/PanelCard/PanelCard.stories.tsx (hunk 3 above)
Comment (roggc) [reply src/Containers/PanelCard/PanelCard.stories.tsx:7]: > Look at Button how it implements actions so you can record button clicks.
> 
> ArgTypes is just
> 
> argTypes: {onClickHandler: action("message on click")}

I tried but it doesn't work

**Comment by ralph-dev** [comment]
> > Order of priority
> 
> > 
> 
> > error Loading Success
> 
> > 
> 
> > Becuase if there's an error who cares if it's loading or success
> 
> > 
> 
> > If there's loading you can be sure it's success
> 
> > 
> 
> > Success is success
> 
> 
> 
> I don't get it

You have a 

If error
Return error component

If loading
Return loading component

Return success component


Order of pretendence

**Comment by ralph-dev** [comment]
if(state.error){
// This is Error block. Do everything related to error hangling.
return <div style={{color:'red'}}> Error state/ every thing related to Error</div>
}
if(state.loading){
// Loading state block. Do everything related to loading show loader/ animation. 
return <div>Loading...</div>
}
// Better and clean code with less ternary conditions.
return <div>
<h3>Display users:</h3>
we have total users : {state.response.length}{state.response.map(user=><User user={user}/>)}
</div>
}

File: src/Containers/PanelCard/PanelCard.stories.tsx (hunk 3 above)
Comment (ralph-dev) [reply src/Containers/PanelCard/PanelCard.stories.tsx:7]: Yes that's all it intends to do, just show you click / responder events that they are working

File: src/Containers/PanelCard/PanelCard.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                <div>
+                    <div>{isLoadingMessage}</div>
+                    <Button onClick={onCancelLoading}>Cancel</Button>
+                </div>
+            );
+        }
+        return <div>{isSuccessMessage}</div>;
+    }, [operationState, isFailureMessage, isSuccessMessage, isLoadingMessage]);
+
+    return <Card {...props}>{renderContent()}</Card>;
```
Comment (ralph-dev) [inline src/Containers/PanelCard/PanelCard.tsx:43]: have in the return statement tree
       <div>
                    <div>{getMessage()}</div>
                    {!!isLoadingMessage && <Button onClick={onCancelLoading}>Cancel</Button>}
                </div>

File: src/Containers/PanelCard/PanelCard.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            return isLoadingMessage;
+        }
+        return isSuccessMessage;
+    }, [operationState, isFailureMessage, isSuccessMessage, isLoadingMessage]);
+
+    return (
+        <Card {...props}>
+            <div>{getMessage()}</div>
+            {operationState.isLoading && (
+                <Button onClick={onCancelLoading}>Cancel</Button>
```
Comment (ralph-dev) [inline src/Containers/PanelCard/PanelCard.tsx:42]: Cancel should not be hardcoded, it is a prop with a default string Cancel.

Ex: cancelComponent = <Button> Cancel </Button>
(Or you can do)
cancelLoadingText = 'Cancel'

File: src/Containers/PanelCard/PanelCard.tsx [hunk 6]
```diff
@@ -0,0 +1,46 @@
+import React, { useCallback } from 'react';
+import { Button } from '@Inputs/Button/Button';
+import { CardProps, Card } from '../Card/Card';
+
+type OperationState = { isFailure: boolean } & { isLoading: boolean };
```
Comment (ralph-dev) [inline src/Containers/PanelCard/PanelCard.tsx:5]: Remove this. This is intersection not union. Also you don't need either just delete the operationState entirely....

        if (operationState.isFailure) {
            return isFailureMessage;
        }
        if (operationState.isLoading) {
            return isLoadingMessage;
        }
        return isSuccessMessage;


Replace the getMessage function with 

return isFailureMessage || isSuccessMessage || isLoadingMessage

**Review by ralph-dev — COMMENTED** [review]
Almost done

**Review by ralph-dev — APPROVED** [review]
Lgtm let's merge

**Comment by roggc** [comment]
@ralph-dev , what about the checks that failed?

**Comment by ralph-dev** [comment]
Don't worry the main check passes

**Pull request merged**